lldb/test/Shell/helper/stepper.py

```diff
@@ -0,0 +1,213 @@
+    return True
+
+
+def __lldb_init_module(dbg, internal_dict):
+    parse_options()
+
+    # Make each debugger command synchronous.
+    dbg.SetAsync(False)
+
+    # Run the program and stop it when it reaches main().
+    if RUN_THE_PROCESS:
+        doit(dbg, 'breakpoint set -n main')
+        doit(dbg, 'run')
+
+    # Step through the program until it exits.
+    gen = 0
+    target = dbg.GetSelectedTarget()
+    process = target.GetProcess()
+    visited_pc_counts = defaultdict(int)
+    while True:
+        gen += 1
+        print(':: Generation', gen)
+
+        thread = process.GetSelectedThread()
+        frame = thread.GetSelectedFrame()
+
+        do_inspection = True
+
+        # Sometimes, lldb gets lost after stepping. This is rdar://70546777.
+        # Try to remind lldb where it is by running 'frame select'.
+        if not frame.GetFunctionName():
+            doit(dbg, 'frame select')
+
+        # Skip frames without valid line entries.
+        line_entry = frame.GetLineEntry()
+        if do_inspection and not line_entry.IsValid():
+            do_inspection = False
+
+        # Skip optimized frames if asked to do so.
+        if do_inspection and ONLY_VISIT_UNOPTIMIZED and \
+                str(frame).endswith(' [opt]'):
+            do_inspection = False
+
+        # Skip non-Swift frames if asked to do so.
+        skip_inspection_due_to_frame_lang = False
+        if do_inspection and ONLY_VISIT_SWIFT and \
+                frame.GuessLanguage() != lldb.eLanguageTypeSwift:
+            do_inspection = False
+            skip_inspection_due_to_frame_lang = True
+
+        if SKIP_N_FRAMES_BETWEEN_INSPECTIONS > 0 and \
+                gen % SKIP_N_FRAMES_BETWEEN_INSPECTIONS != 0:
+            do_inspection = False
+
+        # Don't inspect the same PC twice. Some version of this is needed to
+        # make speedy progress on programs containing loops or recursion. The
+        # tradeoff is that we lose test coverage (the objects visible at this
+        # PC may change over time).
+        cur_pc = frame.GetPC()
+        visit_count = visited_pc_counts[cur_pc]
+        visited_pc_counts[cur_pc] += 1
+        if do_inspection and visit_count > 0:
+            do_inspection = False
+
+        # Inspect the current frame if permitted to do so.
+        if do_inspection:
+            doit(dbg, 'bt')
+
+            # Exercise `frame variable`.
+            if not SKIP_FRAMEVAR:
+                doit(dbg, 'frame variable')
+
+            # Exercise `po`.
+            if not SKIP_PO:
+                get_args = True
+                get_locals = True
+                get_statics = True
+                get_in_scope_only = True
+                variables = frame.GetVariables(get_args, get_locals,
+                                               get_statics, get_in_scope_only)
+                for var in variables:
+                    name = var.GetName()
+                    if not var.GetLocation():
```

Comment:
maybe a comment on why a variable is skipped when it has no location?

Reply:
will add a comment